Dockerfile

```diff
@@ -1,82 +1,63 @@
-COPY --from=dependencies ./deps/node_modules/govuk-frontend/govuk/assets $INSTALL_PATH/app/assets
+RUN mkdir -p $INSTALL_PATH \
+      mkdir -p $INSTALL_PATH/log \
+      mkdir -p $INSTALL_PATH/tmp
 
 WORKDIR $INSTALL_PATH
+COPY . $INSTALL_PATH
 
-# set rails environment
 ARG RAILS_ENV
 ENV RAILS_ENV=${RAILS_ENV:-production}
 ENV RACK_ENV=${RAILS_ENV:-production}
 
-COPY Gemfile $INSTALL_PATH/Gemfile
-COPY Gemfile.lock $INSTALL_PATH/Gemfile.lock
-
-RUN gem install bundler
-
-# bundle ruby gems based on the current environment, default to production
-RUN echo $RAILS_ENV
-RUN \
-  if [ "$RAILS_ENV" = "development" ] || [ "$RAILS_ENV" = "test" ]; then \
-    bundle install --retry 10; \
-  else \
-    bundle install --without development test --retry 10; \
-  fi
-
-RUN mkdir -p $INSTALL_PATH/log
-RUN mkdir -p $INSTALL_PATH/tmp
-
-COPY .rspec $INSTALL_PATH/.rspec
-COPY .rubocop.yml $INSTALL_PATH/.rubocop.yml
-COPY config.ru $INSTALL_PATH/config.ru
-COPY Rakefile $INSTALL_PATH/Rakefile
-
-COPY public $INSTALL_PATH/public
-COPY vendor $INSTALL_PATH/vendor
-COPY lib $INSTALL_PATH/lib
-COPY bin $INSTALL_PATH/bin
-COPY config $INSTALL_PATH/config
-COPY db $INSTALL_PATH/db
-COPY spec $INSTALL_PATH/spec
-COPY app $INSTALL_PATH/app
-
-RUN \
-  if [ ! "$RAILS_ENV" = "development" ] && [ ! "$RAILS_ENV" = "test" ]; then \
-    RAILS_ENV=production bundle exec rake DATABASE_URL=postgresql:does_not_exist --quiet assets:precompile; \
-  fi
-
-COPY ./docker-entrypoint.sh /
-RUN chmod +x /docker-entrypoint.sh
+RUN apk add --update --virtual build-dependencies \
+      build-base \
+      gcc \
+      git \
+      postgresql-dev && \
+      apk add --update --no-cache \
+      bash \
+      libpq \
+      npm \
+      openssl \
+      tzdata && \
+      gem install bundler -v 1.17.3
+
+RUN if [ "$RAILS_ENV" != "production" ]; \
+      then \
+      apk add --no-cache \
+      chromium \
+      chromium-chromedriver; \
+      fi
+
+RUN wget "https://github.com/eficode/wait-for/raw/master/wait-for" -O /bin/wait-for && \
+      chmod a+x /bin/wait-for
+
+RUN if [ "$RAILS_ENV" = "development" ] || [ "$RAILS_ENV" = "test" ]; \
+      then \
+      bundle install --retry 10 --jobs 4; \
+      else \
+      bundle install --without development test --retry 10 --jobs 4; \
+      RAILS_ENV=production bundle exec rake DATABASE_URL=postgresql:does_not_exist --quiet assets:precompile; \
+      fi
+
+RUN npm install --only=production && \
+      apk del build-dependencies
+
+RUN openssl req \
+      -x509 \
+      -nodes \
+      -days 365 \
+      -subj "/C=GB/ST=GB/L=London/O=UK DfE/OU=Digital/CN=default" \
+      -addext "subjectAltName=DNS:mydomain.com" \
+      -newkey rsa:2048 \
+      -keyout /etc/ssl/selfsigned.key \
+      -out /etc/ssl/selfsigned.crt;
+
+ENTRYPOINT ["./docker-entrypoint.sh"]
 EXPOSE 3000
-
-ENTRYPOINT ["/docker-entrypoint.sh"]
-CMD ["rails", "server"]
+CMD sh -c "wait-for db:5432 -- rails server"
```

Comment:
I don't feel good about using `wait-for` in a production environment.  If the database died, I'd prefer the user to see the rails 500 page, plus the service should recover more quickly once the db was back.

Reply:
Thank you for the feedback :)

`wait-for` only applies at container startup and, in its default configuration, introduces a 15 second timeout (counted down at 1 second intervals) to allow the database container to start and accept connections.  Once the rails server is running it will error as expected if the db connection disappears.  `wait-for`'s timeout duration is configurable, but 15 seconds should be sufficient.

`docker-entrypoint.sh` attempts something similar. However, it only checks that the `DATABASE_URL` environment variable is set, not that postgres is actually accepting connections.  It would be sensible to refactor it to use `wait-for` as well although I hadn't gotten to that yet.